cpp/src/arrow/csv/reader.cc

```diff
@@ -545,9 +546,10 @@ class BlockDecodingOperator {
 /////////////////////////////////////////////////////////////////////////
 // Base class for common functionality
 
+template <typename T = std::shared_ptr<io::InputStream>>
```

Comment:
I'm not sure this makes sense as a default parameter.  A reader probably wouldn't intuitively guess this.  Let's just force it to be specified.